cmd/bcpd/app/app_test.go

```diff
@@ -296,36 +226,102 @@ func TestApp(t *testing.T) {
 
+	// create a wallet at contractAddr
+	msg = &cash.SendMsg{
+		Src:  addr,
+		Dest: contractAddr,
+		Amount: &x.Coin{
+			Whole:  2000,
+			Ticker: "ETH",
+		},
+		Memo: "New wallet controlled by a contract",
+	}
+	tx = &Tx{
+		Sum: &Tx_SendMsg{msg},
+	}
+	_, cres = signAndCommit(t, myApp, tx, []Signer{{pk, 2}}, chainID, 4)
+	assert.NotEmpty(t, cres.Data)
+
 	// build and sign a transaction with contract
 	msg = &cash.SendMsg{
 		Src:  contractAddr,
 		Dest: addr2,
 		Amount: &x.Coin{
-			Whole:  2000,
+			Whole:  1000,
 			Ticker: "ETH",
 		},
-		Memo: "Have a great trip!",
+		Memo: "Gift from a contract!",
 	}
 	tx = &Tx{
 		Sum:      &Tx_SendMsg{msg},
 		Multisig: contractID,
 	}
-	sig1, err = sigs.SignTx(pk, tx, chainID, 1)
-	require.NoError(t, err)
-	sig2, err := sigs.SignTx(pk2, tx, chainID, 0)
-	require.NoError(t, err)
-	tx.Signatures = []*sigs.StdSignature{sig1, sig2}
-	txBytes, err = tx.Marshal()
+	_, cres = signAndCommit(t, myApp, tx, []Signer{{ck1, 0}, {ck2, 0}}, chainID, 5)
+	assert.NotEmpty(t, cres.Data)
+
+	// make sure money arrived safely
+	contractAddrKey := namecoin.NewWalletBucket().DBKey(contractAddr)
+	cwquery := abci.RequestQuery{
+		Path: "/",
+		Data: contractAddrKey,
+	}
+	cwqres := myApp.Query(cwquery)
+	checkWalletQuery(t, cwqres, "", 1, map[int]int64{0: 1000}, map[int]string{0: "ETH"})
+}
+
+type Signer struct {
+	pk    *crypto.PrivateKey
+	nonce int64
+}
+
+// signAndCommit signs tx with signatures from signers and submits to the chain
+// asserts and fails the test in case of errors during the process
+func signAndCommit(t *testing.T, app app.BaseApp, tx *Tx, signers []Signer, chainID string, blockHeight int64) (abci.ResponseDeliverTx, abci.ResponseCommit) {
+	for _, signer := range signers {
+		sig, err := sigs.SignTx(signer.pk, tx, chainID, signer.nonce)
+		require.NoError(t, err)
+		tx.Signatures = append(tx.Signatures, sig)
+	}
+
+	txBytes, err := tx.Marshal()
 	require.NoError(t, err)
 	require.NotEmpty(t, txBytes)
 
 	// Submit to the chain
-	header = abci.Header{Height: 4}
-	myApp.BeginBlock(abci.RequestBeginBlock{Header: header})
+	header := abci.Header{Height: blockHeight}
+	app.BeginBlock(abci.RequestBeginBlock{Header: header})
 	// check and deliver must pass
-	chres = myApp.CheckTx(txBytes)
+	chres := app.CheckTx(txBytes)
 	require.Equal(t, uint32(0), chres.Code, chres.Log)
-	dres = myApp.DeliverTx(txBytes)
+
+	dres := app.DeliverTx(txBytes)
 	require.Equal(t, uint32(0), dres.Code, dres.Log)
 
+	app.EndBlock(abci.RequestEndBlock{})
+	cres := app.Commit()
+	return dres, cres
+}
+
+// checkWalletQuery checks the results of a wallet query along with the received wallet
+// maps are used to the wallet state eg. {0: 50000}, {1:"FRNK"} would assert that the first coin whole is 50000 and
+// the second coin ticker is "ETH" in a wallet of at least 2 coins
+func checkWalletQuery(t *testing.T, res abci.ResponseQuery, name string, nbCoins int, wholes map[int]int64, tickers map[int]string) {
```

Comment:
Looks good.
Maybe we can do a better query than sampling a whole and ticker, but this is fine.